velox/functions/sparksql/Factorial.cpp

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/functions/sparksql/Factorial.h"
+#include "velox/expression/ConstantExpr.h"
+#include "velox/expression/VectorFunction.h"
+#include <limits>
+#include <iostream>
+
+namespace facebook::velox::functions::sparksql {
+
+namespace {
+
+/**
+ * Computes the factorial of integers in the range [0...20]
+ *
+ * Returns NULL for inputs which are outside the range [0...20].
+ * Leverages a lookup table for O(1) computation, similar to Spark JVM.
+ */
+class Factorial : public exec::VectorFunction {
```

Comment:
> When implementing a function, simple function is preferred unless the implementation of vector function provides a significant performance gain which can be demonstrated with a benchmark.

Simple function would be preferred if without a significant performance gain.

https://github.com/facebookincubator/velox/blob/main/CONTRIBUTING.md#adding-scalar-functions